Improve number highlighting

dkearns opened the pull request:
This adds support for `ul` and `lu` integer suffixes and disallows invalid use of `_` in number literals.

I've added some error highlighting for the simple cases of leading and trailing underscores.  I could add more thorough error matching if you wish.

I was going to split integers and reals into different syntax groups and link the latter to the Float highlight group but thought you may have grouped them together intentionally.  This would also make expanding the underscore errror highlighting a bit simpler.

I also removed the octal literal highlighting as that's not supported by C#.  It looks like it's been there since the first versions of this syntax file.

dkearns committed:
Remove octal literal highlighting

C# has never had octal literals.

This wasn't being used anyway as leading zeros are allowed in decimal
integer literals and these match later overriding the octal pattern.

dkearns committed:
Improve number highlighting

Highlight ulong integer type suffixes.

Tighten number patterns to only match legal use of underscore
separators.

nickspoons commented:
Thanks @dkearns.

If you want to split integers and reals up and link reals to `Float` that's fine by me, that sounds good.

I think the error highlighting for `_` looks sensible but as a personal preference I don't like too much syntax-level error highlighting, especially in places where you can't avoid typing it as part of typing a correct value, meaning that the error highlight will flash up while typing. So my inclination would be to not add more error highlighting at this stage. I could probably be argued away from this position if you feel strongly about adding more.

dkearns commented:
I've split the numbers up into csInteger and csReal syntax groups.  I know some users appreciate and utilise the distinction.

I tend to share your views regarding error highlighting.  I really only find it useful for whitespace and unbalanced bracket errors.   but even then it's a feature of debatable value for me.  I also often find it a lot of work to implement well.

So, as you gave me the proverbial inch, I removed it for the misplaced underscores but I'm happy to restore it if you think it does have some value.

nickspoons commented:
Thanks again, @dkearns.

Now I'm getting into nitpicking/bikeshedding territory but ... why `csReal`, instead of `csFloat`? Considering that both the C# type is called `float` (and of course `double` which is short for "double precision floating point") and the standard vim highlight group that it links to is called `Float`, shouldn't we call the new highlight `csFloat`? Is there a good reason for naming it `csReal`?

I decided to check the [docs](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/floating-point-numeric-types) and it's not exactly definitive: the main page title is "Floating-point numeric types" but the next section is called "Real literals", which is what we're dealing with here.

I don't know. It seems to me that we're introducing a new term here which might be non-obvious but I could go either way. But if it's ever going to be discussed, it's now :smile:

dkearns commented:
I was reading the [spec](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/lexical-structure#real-literals) at the time.

I don't think it matters much in this case.  High level languages tend to use real but then the VB spec uses "floating point literal" for the same section.
